src/shared/components/common/totp-modal.tsx

```diff
@@ -0,0 +1,266 @@
+                  >
+                    Click here for your TOTP link
+                  </a>
+                  <div className="mx-auto mt-3 w-50 h-50 text-center">
+                    <strong className="fw-semibold">
+                      or scan this QR code in your authenticator app
+                    </strong>
+                    <img
+                      src={this.state.qrCode}
+                      className="d-block mt-1 mx-auto"
+                      alt="TOTP QR code"
+                    />
+                  </div>
+                </div>
+              )}
+              <form id="totp-form">
+                <label
+                  className="form-label ms-2 mt-4 fw-bold"
+                  id="totp-input-label"
+                  htmlFor="totp-input-0"
+                >
+                  Enter TOTP
+                </label>
+                <div className="d-flex justify-content-between align-items-center p-2">
+                  {Array.from(Array(TOTP_LENGTH).keys()).map(i => (
+                    <input
+                      key={
+                        i /*While using indices as keys is usually bad practice, in this case we don't have to worry about the order of the list items changing.*/
+                      }
+                      type="text"
+                      inputMode="numeric"
+                      autoComplete="one-time-code"
+                      maxLength={1}
+                      value={totp[i] ?? ""}
+                      disabled={totp.length !== i}
+                      aria-labelledby="totp-input-label"
+                      id={`totp-input-${i}`}
+                      className="form-control form-control-lg mx-2 p-1 p-md-2 text-center"
+                      onInput={linkEvent({ modal: this, i }, handleInput)}
+                      onKeyUp={linkEvent({ modal: this, i }, handleKeyUp)}
+                      onPaste={linkEvent(this, handlePaste)}
+                      ref={element => {
+                        this.inputRefs[i] = element;
+                      }}
+                    />
+                  ))}
+                </div>
+              </form>
+            </div>
+            <footer className="modal-footer">
+              <button
+                type="button"
+                className="btn btn-danger"
+                onClick={onClose}
+              >
+                {I18NextService.i18n.t("cancel")}
+              </button>
+            </footer>
+          </div>
+        </div>
+      </div>
+    );
+  }
+
+  clearTotp() {
+    this.setState({ totp: "" });
+  }
+
+  async handleShow() {
+    this.inputRefs[0]?.focus();
+
+    if (this.props.type === "generate") {
+      const { getSVG } = await import("@shortcm/qr-image/lib/svg");
+
+      this.setState({
+        qrCode: URL.createObjectURL(
+          new Blob([(await getSVG(this.props.secretUrl!)).buffer], {
+            type: "image/svg+xml",
+          }),
+        ),
+      });
+    }
+  }
+}
```

Comment:
It looks really great, just wish we didn't have to rely on bootstrap modals, and so much custom dom manipulation.

Reply:
I did a bit of refactoring to make handling the modal a bit more react-like.